[GTK] Implement GTK4 accessibility

GeorgesStavracas opened the pull request:
#### da96990ed73dce54453c76e87a1ec3a648a8d8ee
<pre>
[GTK] Implement GTK4 accessibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=227528">https://bugs.webkit.org/show_bug.cgi?id=227528</a>

Reviewed by Carlos Garcia Campos.

WebKit maintains a complete AT-SPI accessible tree within the web
process, under a separate D-Bus name. This conflicts with GTK4&apos;s
expectations of having instances of GtkAccessible objects in process
that can be built in an internal tree.

GTK4 recently introduced support for bridging out-of-process AT-SPI
accessible trees using a new GtkAtSpiSocket object that implements
GtkAccessible [1]. The availability of this object is guarded by the
GTK_ACCESSIBILITY_ATSPI ifdef.

Use this new socket object from GTK to bridge the web page accessible
tree, and the UI process&apos; one. Mark the textview widget as hidden from
the accessible tree.

Create the GtkAtSpiSocket accessible when the web page reports to be
ready. Inject the socket accessible as the first GtkAccessible child
of WebKitWebViewBase.

[1] <a href="https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/6827">https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/6827</a>

* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseAccessibleGetFirstAccessibleChild):
(webkitWebViewBaseAccessibleInterfaceInit):
(webkitWebViewBaseSetPlugID):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:
* Source/WebKit/UIProcess/gtk/KeyBindingTranslator.cpp:
(WebKit::KeyBindingTranslator::KeyBindingTranslator):
* Source/WebKit/UIProcess/gtk/WebPageProxyGtk.cpp:
(WebKit::WebPageProxy::bindAccessibilityTree):
* Source/WebKit/WebProcess/WebPage/glib/WebPageGLib.cpp:
(WebKit::WebPage::platformInitialize):

Canonical link: <a href="https://commits.webkit.org/274201@main">https://commits.webkit.org/274201@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0200806ec10f18906d941ab2b58ce6c2f630fa50

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38141 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17060 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40423 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40690 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33934 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19784 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14404 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32217 "Build is being retried. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 37 flakes 168 failures 1 missing results; Uploaded test results; 21 flakes 145 failures 1 missing results; compiling; 13 flakes 132 failures 1 missing results; Passed layout tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38715 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14412 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33370 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12548 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12507 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41968 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34629 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38363 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13105 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10793 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36554 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14647 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13521 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4978 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14098 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -300,7 +304,11 @@ struct _WebKitWebViewBasePrivate {
-#if !USE(GTK4)
+#if USE(GTK4)
+#if defined(GTK_ACCESSIBILITY_ATSPI)
+    GRefPtr<GtkAccessible> socketAccessible { nullptr };
```
No need for the nullptr initialization, the constructor of GrefPtr already initializes the pointer to nullptr.

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -376,7 +384,10 @@ struct _WebKitWebViewBasePrivate {
 
 #if USE(GTK4)
-WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_WIDGET)
+static void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface *iface);
```
Omit the parameter and move the *

static void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*);

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -387,6 +398,29 @@ static void webkitWebViewBaseDidExitFullScreen(WebKitWebViewBase*);
+        return GTK_ACCESSIBLE(g_object_ref(webView->priv->socketAccessible.get()));
+#endif
+
+    if (auto *widget = gtk_widget_get_first_child(GTK_WIDGET(webView)))
```
auto* widget

Why the first web view child and not the widget itself? The only children the view can have are js dialogs and inspector web view. I don't know how we handle those, TBH.

GeorgesStavracas replied:
This is how `GtkWidget` implements `GtkAccessible.get_first_accessible_child` - it returns the first child `GtkWidget`. Notice that every `GtkWidget` is a `GtkAccessible`  because `GtkWidget` implements this interface.

What I'm doing here is injecting the GtkAtSpiSocket accessible in the a11y tree of the widget:

```
WebKitWebViewBase
        ↓
GtkAtSpiSocket → GtkWidget → GtkWidget → [ ... ]
```

This is necessary because the default implementation WebKitWebViewBase inherits from GtkWidget, that will always return `gtk_widget_get_first_child()`, and will skip the socket accessible entirely. If there's no socket accessible,  we preserve the exact same behavior that of GtkWidget itself:

```
WebKitWebViewBase
        ↓
    GtkWidget → GtkWidget → [ ... ]
```

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -387,6 +398,29 @@ static void webkitWebViewBaseDidExitFullScreen(WebKitWebViewBase*);
+}
+
+static void
+webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface *iface)
```
GtkAccessibleInterface* iface

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -3440,3 +3474,33 @@ void webkitWebViewBaseCallAfterNextPresentationUpdate(WebKitWebViewBase* webView
+
+    GUniqueOutPtr<GError> error;
+    auto tokens = plugID.split(':');
+    auto busName = makeString(':', tokens[0]).utf8().data();
```
You can't do this, String::utf8() returns a temporary CString, so the data returned is already freed when used below in gtk_at_spi_socket_new. In this case you can use g_strdup_print, for example:

```
GUniquePtr<char> busName(g_strdup_printf(":%s", tokens[0].utf8().data()));
```

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -3440,3 +3474,33 @@ void webkitWebViewBaseCallAfterNextPresentationUpdate(WebKitWebViewBase* webView
+    }
+
+    GUniqueOutPtr<GError> error;
+    auto tokens = plugID.split(':');
```
We should probably check we got at least two tokens, even if it's just an ASSERT.

GeorgesStavracas replied:
Done, is that how ASSERT() is meant to be used?

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -3440,3 +3474,33 @@ void webkitWebViewBaseCallAfterNextPresentationUpdate(WebKitWebViewBase* webView
+    GUniqueOutPtr<GError> error;
+    auto tokens = plugID.split(':');
+    auto busName = makeString(':', tokens[0]).utf8().data();
+    auto objectPath = tokens[1].utf8().data();
```
Same happens here with the temporary, in this case just use tokens[1].utf8().data() directly in gtk_at_spi_socket_new() to avoid copies.

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -3440,3 +3474,33 @@ void webkitWebViewBaseCallAfterNextPresentationUpdate(WebKitWebViewBase* webView
+    priv->socketAccessible = adoptGRef(gtk_at_spi_socket_new(busName, objectPath, &error.outPtr()));
+
+    if (priv->socketAccessible) {
+        auto* widget = gtk_widget_get_first_child(GTK_WIDGET(webViewBase));
+        gtk_accessible_set_accessible_parent(priv->socketAccessible.get(), GTK_ACCESSIBLE(webViewBase), GTK_ACCESSIBLE(widget));
```
Why do we make the socket the parent of the first web view child?

GeorgesStavracas replied:
The function [`gtk_accessible_set_accessible_parent`](https://docs.gtk.org/gtk4/method.Accessible.set_accessible_parent.html) is defined as follows:

```c
void
gtk_accessible_set_accessible_parent (
  GtkAccessible* self,
  GtkAccessible* parent,
  GtkAccessible* next_sibling
)
```

What I'm doing here is making the `webViewBase` the parent, and `widget` as the next sibling, of the socket accessible. This is so that GTK4 can iterate over these accessible objects.

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -3440,3 +3474,33 @@ void webkitWebViewBaseCallAfterNextPresentationUpdate(WebKitWebViewBase* webView
+#else
+    UNUSED_PARAM(webViewBase);
+    UNUSED_PARAM(plugID);
+#endif /* defined(GTK_ACCESSIBILITY_ATSPI) */
```
Use C++ style comments

GeorgesStavracas replied:
Done. Is it correct?

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -300,7 +304,11 @@ struct _WebKitWebViewBasePrivate {
     WebHitTestResultData::IsScrollbar mouseIsOverScrollbar;
-#if !USE(GTK4)
+#if USE(GTK4)
+#if defined(GTK_ACCESSIBILITY_ATSPI)
```
Style nit: when there is only one value to check, write `#ifdef foo` rather than `#if defined(foo)`:

```suggestion
#ifdef GTK_ACCESSIBILITY_ATSPI
```

GeorgesStavracas replied:
Done

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -3440,3 +3474,34 @@ void webkitWebViewBaseCallAfterNextPresentationUpdate(WebKitWebViewBase* webView
+    }
+
+    Vector<String> tokens = plugID.split(':');
+    ASSERT(tokens.len() == 2);
```
I would use RELEASE_ASSERT() here to ensure we always crash if the assertion is violated.

Normal ASSERT() is good too, but unfortunately it gets compiled out in release builds.

GeorgesStavracas replied:
Done, thanks

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -3440,3 +3474,34 @@ void webkitWebViewBaseCallAfterNextPresentationUpdate(WebKitWebViewBase* webView
+        auto* widget = gtk_widget_get_first_child(GTK_WIDGET(webViewBase));
+        gtk_accessible_set_accessible_parent(priv->socketAccessible.get(), GTK_ACCESSIBLE(webViewBase), GTK_ACCESSIBLE(widget));
+    } else
+        g_warning("Error creating socket: %s", error->message);
```
A little more info on what is going wrong would be helpful:

```suggestion
        g_warning("Error creating WebKitWebView a11y socket: %s", error->message);
```

carlosgcampos reviewed (approved):
Excellent!

GeorgesStavracas committed:
[GTK] Implement GTK4 accessibility
https://bugs.webkit.org/show_bug.cgi?id=227528

Reviewed by Carlos Garcia Campos.

WebKit maintains a complete AT-SPI accessible tree within the web
process, under a separate D-Bus name. This conflicts with GTK4's
expectations of having instances of GtkAccessible objects in process
that can be built in an internal tree.

GTK4 recently introduced support for bridging out-of-process AT-SPI
accessible trees using a new GtkAtSpiSocket object that implements
GtkAccessible [1]. The availability of this object is guarded by the
GTK_ACCESSIBILITY_ATSPI ifdef.

Use this new socket object from GTK to bridge the web page accessible
tree, and the UI process' one. Mark the textview widget as hidden from
the accessible tree.

Create the GtkAtSpiSocket accessible when the web page reports to be
ready. Inject the socket accessible as the first GtkAccessible child
of WebKitWebViewBase.

[1] https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/6827

* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseAccessibleGetFirstAccessibleChild):
(webkitWebViewBaseAccessibleInterfaceInit):
(webkitWebViewBaseSetPlugID):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:
* Source/WebKit/UIProcess/gtk/KeyBindingTranslator.cpp:
(WebKit::KeyBindingTranslator::KeyBindingTranslator):
* Source/WebKit/UIProcess/gtk/WebPageProxyGtk.cpp:
(WebKit::WebPageProxy::bindAccessibilityTree):
* Source/WebKit/WebProcess/WebPage/glib/WebPageGLib.cpp:
(WebKit::WebPage::platformInitialize):

Canonical link: https://commits.webkit.org/274201@main
